archivist/src/main/scala/uk/ac/wellcome/platform/archive/archivist/models/ArchiveItemJob.scala

```diff
@@ -2,17 +2,53 @@ package uk.ac.wellcome.platform.archive.archivist.models
 
 import uk.ac.wellcome.platform.archive.common.models.bagit.{
   BagItemLocation,
-  BagItemPath
+  BagItemPath,
+  BagLocation
 }
 import uk.ac.wellcome.storage.ObjectLocation
 
-case class ArchiveItemJob(archiveJob: ArchiveJob, bagItemPath: BagItemPath) {
-  def uploadLocation: ObjectLocation = {
-    val bagItemLocation = BagItemLocation(
-      bagLocation = archiveJob.bagUploadLocation,
-      bagItemPath = bagItemPath
+trait ArchiveItemJob {
+  val archiveJob: ArchiveJob
+  val zipEntryPointer: ZipEntryPointer
+  val uploadLocation: ObjectLocation
+}
+
+case class TagManifestItemJob(
+  archiveJob: ArchiveJob,
+  zipEntryPointer: ZipEntryPointer,
+  uploadLocation: ObjectLocation
+) extends ArchiveItemJob
+
+case class DigestItemJob(
+  archiveJob: ArchiveJob,
+  zipEntryPointer: ZipEntryPointer,
+  uploadLocation: ObjectLocation,
+  digest: String
+) extends ArchiveItemJob
+
+object UploadLocationBuilder {
+  def create(bagUploadLocation: BagLocation,
+             itemPathInZip: String,
+             maybeBagRootPathInZip: Option[String] = None): ObjectLocation = {
+    val bagItemUploadLocation = BagItemLocation(
+      bagLocation = bagUploadLocation,
+      bagItemPath =
+        BagItemPath(uploadItemPath(maybeBagRootPathInZip, itemPathInZip))
     )
+    bagItemUploadLocation.objectLocation
+  }
+
+  private def uploadItemPath(maybeBagRootPathInZip: Option[String],
+                             zipPath: String): String =
+    maybeBagRootPathInZip match {
+      case Some(bagRootPathInZip) => stripPrefix(bagRootPathInZip, zipPath)
```

Comment:
I think the standard library can help us here – this could be rewritten

```suggestion
      case Some(bagRootPathInZip) => zipPath.stripPrefix(bagRootPathInZip)
```